torch/csrc/dynamo/guards.cpp

```diff
@@ -674,6 +674,621 @@ static struct PyModuleDef _module = {
+    std::vector<GuardManager*> ret;
+    for (const auto& accessor : _accessors) {
+      ret.emplace_back(accessor->get_guard_manager().get());
+    }
+    return ret;
+  }
+
+  // DEBUG function - Returning raw pointers because we can't return unique_ptr
+  // and pybind does not accept a unique_ptr reference return type.
+  std::vector<LeafGuard*> get_leaf_guards() const {
+    std::vector<LeafGuard*> ret;
+    for (const auto& guard : _leaf_guards) {
+      ret.push_back(guard.get());
+    }
+    return ret;
+  }
+
+ protected:
+  // Keeps a count of how many times this guard manager check function returns
+  // False. This is used for sorting optimization.
+  int64_t _fail_count{0};
+
+ private:
+  // Root of the guard manager, this is the used to install the relational
+  // guard resetters.
+  RootGuardManager* _root;
+
+  // Leaf guards are the terminal guards on this object, e.g, type check on a
+  // list. These guards have to be run before any children are run.
+  //
+  // These leaf guards are not shufflable. In almost all cases, these guards
+  // will have an order, e,g., type(x) is int guard and x == 5 guard. We also
+  // expect very few leaf guards per GuardManager node.
+  //
+  // NB: Why are leaf guards shared ptr? This is primarily to enable relational
+  // guards like `tensor X is not tensor Y`. These guards require multiple
+  // values. We handle it by creating one guard object that holds state and this
+  // guard is installed in many guard managers, hence a shared ptr.
+  std::vector<std::shared_ptr<LeafGuard>> _leaf_guards;
+
+  // GuardAccessors nodes to access the child guards. These guards are
+  // shufflable. On a guard failure, they are sorted based on their fail count
+  // to enable fail fast for the next check.
+  std::vector<std::unique_ptr<GuardAccessor>> _accessors;
+};
+
+/**
+ * RootGuardManager is the root of the guard tree. This is primarily
+ * constructed to hold the relational guard pointers so that we can reset the
+ * state of those guards on guard failure. All the other important
+ * implementation is in GuardManager class.
+ */
+
+class RootGuardManager : public GuardManager {
+ public:
+  // This is the root node, set its _root member to nullptr
+  RootGuardManager() : GuardManager(this) {}
+
+  // Adds the relational guard resetter
+  void add_relational_guard_resetter(
+      std::shared_ptr<RelationalGuard> relational_guard) {
+    _relational_guard_resetters.emplace_back(std::move(relational_guard));
+  }
+
+  // Python visible API to check guard function.
+  bool check(py::handle value) {
+    return check_nopybind(value.ptr());
+  }
+
+  // Python visible API to check_verbose guard function.
+  GuardDebugInfo check_verbose(py::handle value) {
+    return check_verbose_nopybind(value.ptr());
+  }
+
+  // Fast check function.
+  virtual bool check_nopybind(PyObject* value) override { // borrowed ref
+    // reset_state is not thread safe. So acquire a lock.
+    static std::mutex lock;
```

Comment:
This lock:
1) Needs to be shared with check_verbose -- currently we have two locks
2) Can be a class member, since it is fine to run different guard checks in parallel
3) Need to think about interactions with the GIL.  Might need to add:
```
Py_BEGIN_ALLOW_THREADS
std::lock_guard<std::mutex> lock_guard(lock);
Py_END_ALLOW_THREADS
```
That way if this thread blocks waiting we aren't holding the GIL.  Without this I think we could deadlock if you hold the GIL while waiting for the lock.

Reply:
1. Made the mutex a class member.
2. I read about the `Py_BEGIN_ALLOW_THREADS` and your solution makes sense to me. I have not worked with GIL before, so I am not sure if there are other things I need to worry about. I added your suggestion here. (I had to add ; at the end of macros, because clang format was not working, even after disabling them)

Reply:
On second though, using `Py_BEGIN/END_ALLOW_THREADS` might be wrong because they will put lock_guard in braces {}, and lock_guard will be deleted even before we call the critical part. Maybe we should use `Py_BLOCK_THREADS` and ` Py_UNBLOCK_THREADS`

Reply:
Yeah, I didn't look at what the macros do.  What you want to do is release the GIL while waiting for the lock.

Reply:
Ok, I am convinced that the deadlock you mention can happen. I added a note as well to explain how the deadlock could happen. And yes, releasing and acquiring GIL before the `lock_guard` will avoid the deadlock.